SmartAudio Protocol Support

andyp1per opened the pull request:
A fork of https://github.com/ArduPilot/ardupilot/pull/15174 to support SmartAudio

It turns out that a lot of the stability problems are due to issues with our half-duplex implementation. In order to make SmartAudio work properly I have also had to fix half-duplex.

To use:

- Connect the TX pin of a UART to your SmartAudio pin
- Set ```SERIALn_PROTOCOL = 37``` where ```n``` is the UART number
- Set ```SERIALn_OPTIONS=4```
- Set ```SERIALn_BAUD=4800```
- Set ```RCn_OPTION = 94``` to control VTX power from an AUX switch - low means pitmode
- Settings can be manipulated by setting:
```
VTX_OPTIONS - @Bitmask: 0:Pitmode:1:Pitmode until armed,2:Pitmode when disarmed,3:Unlocked
VTX_BAND - @Values: 0:Band A,1:Band B,2:Band E,3:Airwave,4:RaceBand,5:Low RaceBand
VTX_CHANNEL - @Range: 0 7
VTX_FREQUENCY
VTX_POWER - @Description: Video Transmitter Power Level. Different VTXs support different power levels, the power level chosen will be rounded down to the nearest supported power level
```
These settings can also be configured via Spektrum VTX commands, CRSF VTX commands and the OSD parameter support

To work properly, https://github.com/ArduPilot/ardupilot/pull/15549 is also required

skorokithakis commented:
This is great, thank you! Can I ask how this feature will be exposed to end users (parameter-wise)?

squilter commented:
Looking forward to this one. Happy to test, if needed.

andyp1per commented:
> This is great, thank you! Can I ask how this feature will be exposed to end users (parameter-wise)?

I'll wire it into the existing VTX_ parameters that are used by CRSF

skorokithakis commented:
> I'll wire it into the existing VTX_ parameters that are used by CRSF

Does this mean it'll only be usable by CSRF?

andyp1per commented:
> > I'll wire it into the existing VTX_ parameters that are used by CRSF
> 
> Does this mean it'll only be usable by CSRF?

No, the settings are generic - CRSF is the only backend currently, SmartAudio would be the second backend

SergeyBokhantsev commented:
>No, the settings are generic - CRSF is the only backend currently, SmartAudio would be the second backend
@andyp1per did I understand the relation right? :
(CRSF or whatever) <--> AP_VideoTX <--> AP_SmartAudio

So the AP_VideoTX is kind of standardized DTO (and also vtx parameters keeper) between a VTX consumer and a VTX protocol implementation.
And the "CRSF or whatever" may be even not a telemetry, but some internal logic or scripting binding?

I'm trying to understand what should be the model to automatically manage VTX power depending on distance to home etc.

andyp1per commented:
@SergeyBokhantsev yes that's correct. Here's another example - the Spektrum RX decoder can interpret VTX modification packets and insert the data into AP_VideoTX, the connected backend will then make the appropriate changes to the VTX hardware

andyp1per commented:
Ok, this seems to be working pretty well and I have cleaned up the code a lot and made it conform to AP standards. It would be great if some people could give it a try. I have only tested on SmartAudio 2.1 as yet.

skorokithakis commented:
I'd be very happy to test it, but I'd need some details on how, as I'm currently not sure how it can be used.

andyp1per commented:
I'll put some details in the top comment - are you able to build?

skorokithakis commented:
Last time I tried I got some errors, but I should be able to, at least with the Docker image.

skorokithakis commented:
@shellixyz might also be interested in this.

skorokithakis commented:
Just to confirm, SmartAudio can't be used in flight to change transmit power unless you have CSRF or Spektrum?

andyp1per commented:
You can use any mechanism you like to change the VTX_* settings - they are just regular parameters - and the VTX will be updated - I was just giving some examples. You could use a groundstation, or OSD or whatever.

skorokithakis commented:
Hm, right, but all I have available when flying are switches/channels (I can't use the OSD in flight, I'm fairly sure).

Unfortunately, all of my planes that support SmartAudio need to be able to change power in flight, so I don't run AP on those to test the feature...

andyp1per commented:
What behaviour are you looking for? We could easily add an RC function - probably in a different PR though

skorokithakis commented:
@andyp1per just some way to change the power via a channel (changing the frequency is fine to do through the OSD on the ground).

My use case (and that of most people I fly with) is that, when the craft is on the ground/bench/close by, we always start with a low power ("starting power" is a parameter in INAV, we usually set it to the first or second power level, whatever those are.

Then, in flight, I have INAV configured to switch up one power level if a channel goes above 1660 and down if it goes below 1330. That way, I can set a 3-way switch (I use a trim, which my radio has many of) to increase/decrease power (and there's an OSD element that displays power and frequency). This would be the ideal.

Otherwise, a pot could be used in a "set power to whatever quartile the pot is in". This is good too, as long as I can change power in flight with channels, I can figure out the radio programming to send the right thing.

andyp1per commented:
So basically if you could set a switch or pot so that PWM_MIN = PITMODE and PWM_MAX = MAX POWER?

skorokithakis commented:
Not exactly, AFAIK SmartAudio supports 4 levels, so that's what I switch. It starts at 1 or 2, hitting the switch up once goes to 3, again to 4, one down to 3 again, etc. I generally don't run at max power unless I go very far.

Splitting a channel into 4 regions and setting power according to what quartile the channel is in will also work fine, though. So, for example, if the channel were at 1600, SA would switch to power level 3.

andyp1per commented:
Generally our switches are stateless, so the second approach is better - that's essential what I meant. I believe later versions of some VTX's allow power to be set precisely in dbm so that would also work well.

skorokithakis commented:
Yeah, understandable. I can certainly use one of the spare pots for setting power, I think that will be fine for most people who are looking forward to this functionality as well.

TunaLobster commented:
Tagging a few issues that are relevant to this PR.
#9565
#10084
#14383

squilter commented:
Tried to get it running today, but ran into some issues. No signal out of the pin on a logic analyzer. Possibly a problem with my hwdef. Will report back if I get it working.

andyp1per commented:
Thanks for trying. You are using the *TX* pin of your UART, yes?

squilter commented:
I am using the tx pin. I had to add this to my hwdef, so I suspect that I missed something. Can you take a quick look:
https://github.com/squilter/ardupilot/commit/0d8ab9aefe3ffaf33a49f32cf9e77e489177fbd0

Proof that pin PA2 is correct is here:
https://store.myairbot.com//media/iverve/uploadpdf/1527195262_OMNIBUSF4NanoV6.pdf

squilter commented:
Got it working on an old pixhawk. I will have to get it working on my newer flight controller in order to test more thoroughly.

squilter commented on libraries/AP_SmartAudio/AP_SmartAudio.h:
```diff
@@ -0,0 +1,267 @@
+#include <AP_HAL/AP_HAL.h>
+
+#ifndef HAL_SMARTAUDIO_ENABLED
+#define HAL_SMARTAUDIO_ENABLED !HAL_MINIMIZE_FEATURES
```
Is it possible to warn the user if they try to enable smartaudio but it isn't included in the build? It is often obvious to the user that some feature isn't included because relevant parameters don't exist. But in this case, all the VTX_* parameters will exist because they are also used for crossfire.

andyp1per replied:
The obvious thing to do is remove the serial protocol number - but we don't do that for any other protocol, so what is there now is at least consistent :)

squilter commented:
I have resolved my issue and was able to test. Sorry for clogging the comments with this unrelated issue, but here is the solution in case anybody ever stumbles upon this thread: Revision 6.1 of the Omnibus F4 Nano has an undocumented errata (erratum?) where pin PA2 does not actually trace to the pad where it supposedly is supposed to trace to.

squilter commented on libraries/AP_SmartAudio/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,578 @@
+            }
+        }
+
+        debug("update_params(): freq %d->%d, chan: %d->%d, band: %d->%d, pwr: %d->%d, opts: %d->%d",
```
This is triggering every 5 seconds for me, even though the conditions above seem to suggest that it should only happen when there is a change. Is this intentional?

andyp1per replied:
Yeah, so this is one of the issues I have to resolve. If you pick a combination of settings that aren't allowed by the VTX it will keep trying to set them even though the VTX has replied with something different to what you set. Common ones are picking a frequency that is only allowed if you unlock.

andyp1per replied:
Fixed now I believe

squilter commented:
I am unable to get to pit mode by setting the first bit of the VTX_OPTIONS bitmask. Instead, I am able to get to pit mode by setting VTX_POWER to 0.

Although I have not used crossfire vtx, I believe that this behavior is inconsistent between the two drivers. This inconsistency might get ugly later on if we add new features like "always use pit mode when disarmed". (coincidentally this is a feature that I am keen to add after this gets merged).

andyp1per commented:
What VTX do you have and what version of SmartAudio does it use?

squilter commented:
I am using a TBS Unify Pro32 Nano 5G8 V1.1. The documentation says that it is using SmartAudio 2.1.

andyp1per commented:
Ok, what VTX settings are you using? I have one of those so will be able to reproduce.

squilter commented:
Maybe this behavior is actually correct. Page 12 of [the manual](https://www.team-blacksheep.com/tbs-unify-pro32-manual.pdf) says that setting power level to 0 will override the pit mode flag. Let me try it a bit more so that I can understand the behavior better.

squilter commented:
The manual seems to imply that tbs unify devices have non-standard behavior. They describe their nonstandard behavior like this:
setting the pit mode bit will actually enable "pit mode on startup"
setting vtx power to 0 will immediately  enable pit mode

However, the "pit mode on startup" doesn't seem to be working for me. The pit mode bit doesn't appear to have any effect.

andyp1per commented:
@skorokithakis I have added support for using a 6-position switch or pot to control power levels. You can get fewer increments by settting VTX_MAX_POWER. Its controlled through RC_OPTION 94.

skorokithakis commented:
@andyp1per that's fantastic, thank you! I'll try to compile and test soon.

andyp1per commented:
@squilter I have changed the logic so that the VTX parameters are set to what the VTX was actually able to configure - it should mean you avoid the repeated updates. I have also changed the GCS announcement to be broadcast every time a setting is successfully changed.

tridge commented on libraries/AP_RCTelemetry/AP_VideoTX.cpp:
```diff
@@ -61,9 +61,15 @@ const AP_Param::GroupInfo AP_VideoTX::var_info[] = {
     // @Description: Video Transmitter Options.
     // @User: Advanced
-    // @Bitmask: 0:Pitmode
+    // @Bitmask: 0:Pitmode,1:Unlocked
```
add in description "Do not enable the Unlocked option unless you have appropriate permissions in your jurisdiction to transmit at high power levels"
It should also describe what Pitmode is
pitmode on disarm sounds nice

giacomo892 replied:
having pitmode on disarm or just lower power setting is indeed nice, but we should only be sure it goes there only when we power up the first time and not on next disarms (you crash and disarm you want full video power to be able to find your machine for instance), also we need to be sure that power doesn't change in case we watchdog reset since the last thing you want when flying is to loose the control due to wdt and the video from the machine even briefly.

skorokithakis replied:
I'll second this, though I'd point out that it's better if this is an option, as many people fly in groups and would prefer to have pit mode on all disarms, to not disturb other fliers when they land.

andyp1per replied:
I added both options

tridge commented on libraries/AP_VideoTX/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,619 @@
+
+    //bool tried = false;
+
+    while (true) {
```
wait for system initialised?

andyp1per replied:
Fixed

tridge commented on libraries/AP_VideoTX/AP_SmartAudio.cpp:
```diff
@@ -0,0 +1,619 @@
+
+        // nothing going on so give CPU to someone else
+        if (!_is_waiting_response && !_initialised) {
+            hal.scheduler->delay(500);
```
100ms? for human response to a switch

andyp1per replied:
Fixed

peterbarker commented on libraries/AP_RCTelemetry/AP_VideoTX.cpp:
```diff
@@ -59,11 +59,17 @@ const AP_Param::GroupInfo AP_VideoTX::var_info[] = {
+    // @Description: Video Transmitter Options. Pitmode puts the VTX in a low power state. Unlocked enables certain restricted frequencies and power levels. Do not enable the Unlocked option unless you have appropriate permissions in your jurisdiction to transmit at high power levels.
     // @User: Advanced
-    // @Bitmask: 0:Pitmode
+    // @Bitmask: 0:Pitmode:1:Pitmode until armed,2:Pitmode when disarmed,3:Unlocked
```
```suggestion
    // @Bitmask: 0:Pitmode,1:Pitmode until armed,2:Pitmode when disarmed,3:Unlocked
```

andyp1per replied:
Fixed

andyp1per commented:
Remo has confirmed that lock does not work by design. Once you have unlocked a VTX you need to use the manual method to lock again.

squilter commented:
Just tested the new changes to pit mode handling, and pit mode on arm/disarm. Works great!

andyp1per commented:
I've removed the UART changes since they are in a separate PR

skorokithakis commented:
Can you please link the other PR, for those of us who are interested?

andyp1per commented:
It's linked at the top? I don't have permissions to change the sidebars.

skorokithakis commented:
Ah I see it, thank you.

Hwurzburg commented:
Power out..pls add wiki needed label

karambolis8 commented:
Any estimate when it's going to be included in release version?

andyp1per commented:
4.1 beta should start soonish

Target0815 commented:
Is the function of Smartaudio different on different processors?

I have a H743-Wing where the switching through the power levels is displayed in the OSD as well as in real life. 

But on a F765-Wing with the same configuration it does not work. Here only zero or the power, which was activated in the setup of the VTX (default 25 mW), can be switched.

Mind you, the configuration of both FCs is identical. The H743-Wing runs with a Latest from the beginning of April, the F765-Wing with Latest from the end of April.

The VTX used, a TBS Unify Pro32 HV, is unlocked.

andyp1per commented:
They should be the same. Are you using the same VTX? Might be a VTX firmware difference.

Target0815 commented:
Yes, it is the same VTX.

Target0815 commented:
I found a few more things to discuss here:
https://discuss.ardupilot.org/t/smartaudio-wrong-frequency-only-one-power-level/71092
